### PR TITLE
Use Okio for encoding and decoding UTF-8 Strings

### PR DESCRIPTION
**Pull request by MichaelEvans**

### CODE REVIEW

File: serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerInput.java [hunk 1]
```diff
@@ -262,7 +267,11 @@ private long readLongHeader(byte expectedType) throws IOException {
 
     @NotNull
     private String decodeUtf8String(int length) throws IOException {
-        return "";
+        ByteString string = ByteString.of(mByteBuffer.array(),
```
Comment (amandeepg) [inline serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerInput.java]: final

File: serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java [hunk 2]
```diff
@@ -254,70 +257,24 @@ private void ensureCapacity(int sizeNeeded) {
      * Encodes the string into the buffer using UTF-8.
      */
     private void encodeString(@NotNull String string) {
-        final int length = string.length();
-        final int headerPosition = mByteBuffer.position();
-        writeIntHeader(SerializerDefs.TYPE_STRING_ASCII, length);
+        ByteString byteString = ByteString.encodeUtf8(string);
```
Comment (amandeepg) [inline serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java]: final

Comment (cesar1000) [reply serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java]: Also, you should move this into the else below

File: serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java [hunk 3]
```diff
@@ -254,70 +257,24 @@ private void ensureCapacity(int sizeNeeded) {
...
         boolean isAscii = true;
-        for (int i = 0; i < length; ++i) {
+        for (int i = 0; i < string.length(); ++i) {
             final int ch = (int) string.charAt(i);
-            if (ch < 0x80) {
-                ensureCapacity(ByteBufferSerializerDefs.SIZE_BYTE);
-                mByteBuffer.put((byte) ch);
-            } else {
+            if (ch > 0x80) {
```
Comment (cesar1000) [inline serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java]: \>=

File: serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java [hunk 4]
```diff
@@ -254,70 +257,24 @@ private void ensureCapacity(int sizeNeeded) {
...
-        // If the string is not ASCII, update the header.
-        if (!isAscii) {
-            final int currentPosition = mByteBuffer.position();
-            mByteBuffer.position(headerPosition);
-            writeIntHeader(SerializerDefs.TYPE_STRING_UTF8, length);
-            mByteBuffer.position(currentPosition);
+        if (isAscii) {
+            writeIntHeader(SerializerDefs.TYPE_STRING_ASCII, string.length());
+            ensureCapacity(string.length() * ByteBufferSerializerDefs.SIZE_BYTE);
+            mByteBuffer.put(ByteBuffer.wrap(string.getBytes()));
```
Comment (cesar1000) [inline serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java]: This is copying the string into a byte array, right? Can you avoid that?

File: serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java (hunk 2 above)
Comment (cesar1000) [reply serialization/src/main/java/com/twitter/serial/stream/bytebuffer/ByteBufferSerializerOutput.java]: Also, this is copying the string as well. Why are we removing this code at all? Is this code suspect as well?

**Pull request closed without merging**